rfcs/0000-libp2p.md

```diff
@@ -0,0 +1,149 @@
+      val unsubscribe : t -> unit Deferred.Or_error.t
+    end
+
+    (** Get the topic handle for a string.
+      *
+      * You can think of this as hashing the string.
+      *)
+    val topic_of_string : string -> topic Deferred.Or_error.t
+
+    (** Subscribe to a pubsub topic.
+      *
+      * Fails if already subscribed. If it succeeds, incoming messages for that
+      * topic will be written to the pipe.
+      *)
+    val subscribe : t -> topic -> (bytes Envelope.Incoming.t, _, _) Strict_pipe.Writer.t -> Subscription.t Deferred.Or_error.t
+
+    (** Validate messages on a topic with [f] before forwarding them. *)
+    val register_validator : t -> topic -> f:(bytes Envelope.Incoming.t -> bool Deferred.t) -> unit
+  end
+
+  (** Create a libp2p network manager.
+    *
+    * The provided [rpcs] will be used to handle calls to [rpc_peer]>
+    *)
+  val create : me:Keypair.t -> rpcs:Host_and_port.t Rpc.Implementation.t list -> t Deferred.Or_error.t
+
+  (** List of all peers we know about. *)
+  val peers : t -> Network_peer.Peer.t list Deferred.t
+
+  (** Randomly pick a few peers from all the ones we know about. *)
+  val random_peers : t -> int -> Network_peer.Peer.t list Deferred.t
+
+  (** Perform Jane Street RPC with a given peer *)
+  val rpc_peer :
+       t
+    -> Network_peer.Peer.t
+    -> (Versioned_rpc.Connection_with_menu.t -> 'q -> 'r Deferred.Or_error.t)
+    -> 'q
+    -> 'r Deferred.Or_error.t
+
+  (** Try listening on a multiaddr.
+    *
+    * If successful, returns an alternate version of the multiaddr that other
+    * nodes can use to connect. For example, if listening on
+    * ["/ip4/127.0.0.1/tcp/0"], it might return ["/ip4/127.0.0.1/tcp/35647"]
+    * after the OS selects an available listening port.
+    *)
+  val listen_on : t -> Multiaddr.t -> Multiaddr.t Deferred.Or_error.t
+
+  (** Stop listening, close all connections and subscription pipes. *)
+  val shutdown : t -> unit Deferred.Or_error.t
+end
+```
+
+Concretely, this will be implemented by spawning a child process that speaks a simple JSON protocol. This will let us use `go-libp2p` (which seems to be the most robust libp2p implementation at the moment) without figuring out how to get Go and async working in the same process.
+
+`Pubsub` will replace the `Gossip_net` functionality. Topic names will be `{genesis state hash}-{description}`: `{hash}-states`, `{hash}-transaction-pool-diffs`, `{hash}-snark-pool-diffs` . Additional topics are easy to add. The raw pubsub messages are uninterpreted bytes; it will be up to the caller to map the pipe with useful de-serialization.  Topic message validation might be a problem. There's a default timeout of 150ms waiting to validate a message. We can turn it up, but incoming messages can come with their own timeout. There are some synchronous operations that can take longer than 150ms in our daemon (I think sparse ledger creation exceeds that right now, at least when ledger_depth=30).
```

Comment:
Can you speak to if libp2p offers adversary protection? Kademlia for example uses a DHT to determine the topology of the network, which increases the difficulty of an adversary performing an eclipse attack.

Reply:
In their docs they claim: `This DHT is modeled after kademlia with S/Kademlia modifications.` From the implementation and issues, it seems they implement at least 4.4 from the [S/Kademlia paper](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.68.4986&rep=rep1&type=pdf). nodeIDs are also the hashes of the peer IDs (which themselves are either public keys or hashes of public keys). This is what prevents the easiest forms of eclipse attacks. Beyond the more robust routing/lookup algorithm, I don't think there are significant sybil mitigations.